temporal-sdk/src/main/java/io/temporal/internal/sync/CompletablePromiseImpl.java

```diff
@@ -275,9 +280,8 @@ private <U> Promise<U> then(Functions.Proc1<CompletablePromise<U>> proc) {
    * @return true if there were any handlers invoked
    */
   private boolean invokeHandlers() {
-    for (Functions.Proc handler : handlers) {
-      handler.apply();
-    }
+    executor.execute(() -> handlers.forEach(Functions.Proc::apply));
```

Comment:
I believe this change is not backward compatible and can break determinism.

Reply:
@mfateev are you referring to the forEach or the executor deferment?

Reply:
I added documentation to clarify the synchronous requirement and changed to a plain for() loop.

Reply:
My point is that switching from direct call to an executor that might change the order of execution of these calls can lead to non determinism. Such changes should be protected by versioning/patching.

Reply:
Ah, I see.  FWIW: I was careful to design the executor chain in a way that should respect the original order of interceptors, and I retained the original order of the for loop as well.  I suspect your concern would be if someone added an interceptor that creates an NDE situation, but I think that if that were the case, they should follow patching procedures in the intercept code.  I don't think we could solve that at this level.  Thoughts?

Reply:
@mfateev I looked, and I don't see any apparent examples of SDK code using patching internally.  Since I don't believe the PR code changes the ordering, the only NDE potential that I see would be if an interceptor introduced a determinism issue in its handler.  However, that would also concern any existing interception points, so it's unclear if this proposed one should be treated differently.  

This is a blocking issue for me, so I am highly motivated to see it get merged.  So please let me know what you'd like to see done here, and I will do my best to make it happen.